lwtime/reltime.go

```diff
@@ -0,0 +1,211 @@
+// Author:: Salim Afiune Maya (<afiune@lacework.net>)
+// Copyright:: Copyright 2020, Lacework Inc.
+// License:: Apache License, Version 2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package lwtime
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type RelDate struct {
+	Year  int
+	Month time.Month
+	Day   int
+}
+
+func Mondays(year int) (mondays []RelDate) {
+	start := time.Date(year, 1, 1, 0, 0, 0, 0, time.Now().Location())
+	d, _ := time.ParseDuration("24h")
+	startYear := start.Year()
+	for startYear == start.Year() {
+		if start.Weekday() == time.Monday {
+			year, month, day := start.Date()
+			mondays = append(mondays, RelDate{year, month, day})
+		}
+		start = start.Add(d)
+	}
+	return mondays
+}
+
+type RelTimeUnit string
+
+const (
+	relTimeRE             = `^([+-])?(?:(\d+)(\w+))?(?:@(\w+))?$`
+	Year      RelTimeUnit = "y"
+	Month     RelTimeUnit = "mon"
+	Week      RelTimeUnit = "w"
+	Day       RelTimeUnit = "d"
+	Hour      RelTimeUnit = "h"
+	Minute    RelTimeUnit = "m"
+	Second    RelTimeUnit = "s"
+)
+
+func (rtu RelTimeUnit) IsValid() bool {
+	switch RelTimeUnit(strings.ToLower(string(rtu))) { // inline lowercase conversion
+	case Year, Month, Week, Day, Hour, Minute, Second, RelTimeUnit(""):
+		return true
+	}
+	return false
+}
+
+func (rtu RelTimeUnit) SnapTime(inTime time.Time) (outTime time.Time, err error) {
+	// immediately short circuit if snap is invalid
+	if !rtu.IsValid() {
+		err = errors.New(fmt.Sprintf(
+			"snap (%s) is not a valid relative time unit", rtu))
+		return
+	}
+
+	year, month, day := inTime.Date()
+	hour := inTime.Hour()
+	minute := inTime.Minute()
+	second := inTime.Second()
+	nano := inTime.Nanosecond()
+
+	switch RelTimeUnit(strings.ToLower(string(rtu))) {
+	case Week:
+		year, week := inTime.ISOWeek()
+		relDate := Mondays(year)[week-1]
+		outTime = time.Date(
+			relDate.Year,
+			relDate.Month,
+			relDate.Day,
+			0, 0, 0, 0, inTime.Location(),
+		)
+		return
+	case Year:
+		month = 1
+		fallthrough
+	case Month:
+		day = 1
+		fallthrough
+	case Day:
+		hour = 0
+		fallthrough
+	case Hour:
+		minute = 0
+		fallthrough
+	case Minute:
+		second = 0
+		fallthrough
+	case Second:
+		nano = 0
+	}
+	outTime = time.Date(year, month, day, hour, minute, second, nano, inTime.Location())
+	return
+}
+
+type RelTime struct {
+	Num  string
+	INum int
+	Unit RelTimeUnit
+	Snap RelTimeUnit
+}
+
+func (rt *RelTime) Parse(s string) error {
```

Comment:
Why would you want to have a Relative time and after that, Parse another time?

I feel that you need a constructor here again. Let's sync up over Zoom 📹. ⭐

Reply:
Fixed.